#810 - EEZ report generation

andres-arana opened the pull request:
# Introduction

This pull request contains the client-side changes for https://github.com/SkyTruth/pelagos-server/issues/810.
# Summary of changes
- Adds new fields to the cartodb layers, `reportable` and `reportBaseUrl`. When the layer is reportable, a button is generated in the info window which allows the user to start the report generation process
- Formats the schema documentation, adding the changes on the schema for the report generation

andres-arana committed:
Implements reportable layers

This implements a reportable field on layers that enable a "generate
activity report" button on the info pane for that layer. For now, that
button just creates a popup asking for confirmation, but not posting
that anywhere.

redhog commented on js/app/Visualization/UI/LayerReportDialog.js:
```diff
@@ -0,0 +1,66 @@
+
+        // TODO: Complete this once the server endpoint is done, eez should be
+        // replaced by the actual column name.
+        var url = layerData.reportBaseUrl + "/sub/eez=" + eez + "/report";
```
I think that event.data.reportable at https://github.com/SkyTruth/pelagos-client/pull/108/files#diff-38f3be8d788728b2fcdf96410c5e1da6R126 above should contain the report query, and be replaced by 

```
new ObjectTemplate(event.data.reportQuery).eval({polygonValue: SOME_VALUE_FROM_SELECTED_POLYGON})
```

How do you plan to extract / how is the current value for the `eez` variable above calculated?

andres-arana replied:
Agreed! It will certainly be an object template with some value from the selected polygon, to keep it generic (I think that MPA reports will be coming soonish).

However, I'm still not sure about what exactly I will be sending to the backend here. I wanted to talk with @geowurster about how the data will look like to see what will be more convenient to post here, but in the meantime we can merge this as it is and I'll change it once I have the backend ready.

geowurster replied:
@andres-arana What info will be going into the report - is there a ticket anywhere describing the feature?

andres-arana replied:
@geowurster, the ticket is https://github.com/SkyTruth/pelagos-server/issues/810. Basically, what we want is, for a given EEZ, all the ships that were detected inside the EEZ in a period of time, with the date it was first detected in there and the date it was last seen in there.

What I want to know is where the EEZ field is in the new pipeline. I suppose that some task will classify each AIS record and append a field somewhere that indicates if the vessel was in a given EEZ. Is that right? And on the other hand, how do we identify a given EEZ? Do we have a numeric ID for each EEZ, some kind of code, or the name of the EEZ identifies it uniquely?

geowurster replied:
@andres-arana Will likely do something like we did before: a JSON encoded `regions` field with values like `['EEZ:Kiribati', 'MPA:PIPA', 'MPA:Something Else']`.  We can handle the ID mapping in the pipeline and human-readable text seems like the most useful output.

In order to build queries we will need to publish our layers in a location that is easy to browse.  Maybe we could use CartoDB to let us poke around in the layers to find the correct ID.

andres-arana committed:
Fixes invalid [Object object] legend

The selection event data field has a toString method, which surprisingly
doesn't return an actual string, but instead returns a jquery node. This
allows the originator of the event data to attach event handlers to the
node, which is then appended into the info window.

We are not refactoring that method into something else in this branch
though. I created a ticket at #109.
